L1Trigger/L1TTrackMatch/plugins/L1GTTInputProducer.cc

```diff
@@ -0,0 +1,532 @@
+                                         << ptBitsShifted.to_string(2) << " = " << ptBitsShifted.to_uint() << ")\n"
+                                         << "produce::etaBits after all conversions " << etaBits.to_string(2) << " ("
+                                         << etaBitsShifted.to_string(2) << " = " << etaBitsShifted.to_uint() << ")\n"
+                                         << "produce::word after all conversions "
+                                         << vTTTrackOutput->back().getTrackWord().to_string(2);
+    }
+
+    // Fill the remaining outputs
+    vPtOutput->push_back(expectedPt);
+    vEtaOutput->push_back(expectedEta);
+  }
+
+  if (debug_ >= 1) {
+    edm::LogInfo("L1GTTInputProducer") << "\nNumber of converted tracks: " << nOutput << "\n\n"
+                                       << "q/r ==> pt conversion:\n"
+                                       << "\tError Threshold: " << kPTErrThresh << "%\n"
+                                       << "\tMax error: " << maxErrEpsilonPt
+                                       << " GeV difference with percentage: " << maxErrPercPt << "% @ "
+                                       << 100.0 * maxErrEpsilonPt / maxErrPercPt << " GeV"
+                                       << "\n"
+                                       << "\tError @ max range: " << minExpectedPt
+                                       << " GeV with precentage: " << minErrPercPt << "%"
+                                       << "\n"
+                                       << "\tTotal number of errors: " << error_pt_c << "\n\n"
+                                       << "tan(lambda) ==> eta conversion:\n"
+                                       << "\tError Threshold: " << kEtaErrThresh << "\n"
+                                       << "\tMax error: " << maxErrEpsilonEta << " with percentage: " << maxErrPercEta
+                                       << "% @ " << 100.0 * maxErrEpsilonEta / maxErrPercEta << "\n"
+                                       << "\tTotal number of errors: " << error_eta_c;
+  }
+
+  if (error_pt_c + error_eta_c) {
+    edm::LogError("L1GTTInputProducer") << "produce::" << error_pt_c << "/" << error_eta_c
+                                        << " pt/eta mismatches detected!!!";
+  }
+
+  // Put the outputs into the event
+  iEvent.put(std::move(vTTTrackOutput), outputCollectionName_);
+  iEvent.put(std::move(vPtOutput), "L1GTTInputTrackPtExpected");
+  iEvent.put(std::move(vEtaOutput), "L1GTTInputTrackEtaExpected");
+}
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void L1GTTInputProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
+  descriptions.addDefault(desc);
```

Comment:
can you implement this instead of (or in addition to) providing a cfi file ?

Since you already wrote a cfi file, you can convert it to what the `fillDescription` method should be:
```bash
$ edmPythonConfigToCppValidation L1GTTInputProducer_cfi.py
```
will give you
```c++
#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"

void
L1GTTInputProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
  // L1GTTInputProducer
  edm::ParameterSetDescription desc;
  desc.add<int>("debug", 0);
  desc.add<edm::InputTag>("l1TracksInputTag", edm::InputTag("TTTracksFromTrackletEmulation","Level1TTTracks"));
  desc.add<std::string>("outputCollectionName", "Level1TTTracksConverted");
  descriptions.add("L1GTTInputProducer", desc);
  // or use the following to generate the label from the module's C++ type
  //descriptions.addWithDefaultLabel(desc);
}
```

My suggestion would be to use the default label.
You can also add the same comment you have in the python file:
```c++
#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"

void
L1GTTInputProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
  // L1GTTInputProducer
  edm::ParameterSetDescription desc;
  desc.add<int>("debug", 0)->setComment("Verbosity levels: 0, 1, 2, 3");
  desc.add<edm::InputTag>("l1TracksInputTag", edm::InputTag("TTTracksFromTrackletEmulation", "Level1TTTracks"));
  desc.add<std::string>("outputCollectionName", "Level1TTTracksConverted");
  descriptions.addWithDefaultLabel(desc);
}
```

I guess the "default label" for this module should be `l1GTTInputProducer`.